orttraining/orttraining/core/graph/mixed_precision_transformer.cc

```diff
@@ -49,15 +47,13 @@ static const std::unordered_map<std::string, std::vector<int>> stage1_fp32_node_
     {"DropoutGrad", {2}},
 };
 
+// Currently the list here is same as stage1 above due to empty FP32_Nodes.
+// It's possibile we will have more FP32 nodes added, this map will also be extended.
 static const std::unordered_map<std::string, std::vector<int>> stage2_fp32_node_args = {
     {"TrainableDropout", {1}},
     {"TrainableDropoutGrad", {2}},
     {"Dropout", {1}},
     {"DropoutGrad", {2}},
-    {"SparseSoftmaxCrossEntropy", {0, 2}},
-    {"SparseSoftmaxCrossEntropyGrad", {0, 1, 3}},
-    {"SoftmaxCrossEntropyLoss", {0, 2}},
-    {"SoftmaxCrossEntropyLossGrad", {0, 1, 3}},
```

Comment:
Surprised to see loss scale in FP16. It is definitely a bug in the implementation. What does loss subgraph for BERT look like? If loss scale is in FP32, then the output of SparseSoftmaxCrossEntropy should not be casted to FP16.

Reply:
Current logic just add CastF32 right before and CastF16 right after the two SparseSoftmaxCrossEntropy and SparseSoftmaxCrossEntropyGrad node only. All other parts of the graph are in FP16, including those nodes between SparseSoftmaxCrossEntropy and SparseSoftmaxCrossEntropyGrad topologically. So the loss output is FP16, then there is a CastFP32 applied to loss_scale before it times to the FP16 loss.

Reply:
In current implementation, is SparseSoftmaxCrossEntropy followed by the op of loss scale? If so, then if loss scale is in FP32, the the CastFP16 can't be added to the output of SparseSoftmaxCrossEntropy. Otherwise, the two inputs of loss scale op will be different, which is not allowed for ONNX Mul op. Am I right? Or anything I missed?

Reply:
For BERT, because there are two SparseSoftmaxCrossEntropy (mlm_loss and nsp_loss) and there is Add node to add them to get the final loss, so start from Add node, they are in FP16. If the loss is directly from the SparseSoftmaxCrossEntropy, then loss and loss_scale will remain FP32.